pandas/_libs/tslibs/src/datetime/np_datetime_strings.c

```diff
@@ -66,10 +66,25 @@ This file implements string parsing and creation for NumPy datetime.
  *
  * Returns 0 on success, -1 on failure.
  */
+
+inline int format_startswith(char ch, int format_len, char format, int exact) {
```

Comment:
Still not sure about this function. What is the advantage of this over using `strncmp`? Seems like it would be cleaner to just use that plus some combination of `exact` outside of this

Reply:
You might also be interested in the POSIX `strstr` function for substring searches with the `exact` argument

Reply:
OK I think I understand a bit more now what you are trying to do. I think it would be best if you just kept a local variable for characters consumed, which you can increment every time you increment the format pointer (you are kind of doing this anyway).  You don't really need a function but can rather just do `if (format++ != '%') { // handle error }`

The order of operations will do the postfix increment after assignment, so this consolidates what you are trying to do

Reply:
Just checking I've understood, is the suggestion to get rid of the function and each time do something like
```
if (((format_len > 0) && (format++ != '%')) || (exact && (format_len<=0))){
    if (format_len > 0) format_len--;
    goto parse_error;
}
```
?

I presume I've misunderstood because to me this looks more complicated and loses the docstring 😄

Reply:
It is possible to even do everything in one single if statement, but it then starts to look as one of these tricky C language questions... This separate function has comments, it is more clear imho, and is very likely to get inlined.

Reply:
Its not a matter of inlining as much as code organization. For instance, this block:

```c
    if (!format_startswith('%', format_len, *format, exact)) goto parse_error;
    if (format_len) {++format; --format_len;}
    if (!format_startswith('Y', format_len, *format, exact)) goto parse_error;
    if (format_len) {++format; --format_len;}
```

Would seem more naturally expressed as:

```c
if (format_len < 2 && exact) {
  goto parse_error;
} 

if (*format++ != '%') {
  goto parse_error;
} 

if (*format++ != 'Y') {
  goto parse_error;
}
```

There's some code golf you can do therein to shorten it but that's not really what I'm after. Moreso we should refactor this to avoid trying to cram a lot into a function with a lot of side effects, as we don't use that type of pattern much if at all in our code base, and a lot of the character by character checks being done are overkill

Reply:
The other advantage of this is you _could_ provide an actual error message as to what went wrong at what position of the format, though not critical to scope for this PR

Reply:
I agree. Can't we just refactor everything and replace it with something like

1) match given format with the full format and return position of mismatch/success that depends on lengths and exactness.
2) Parse the string with sprintf or something like this, with additional validity checks. Provided that 1 is a success
?
Unless doing everything in one go is a performance decision.

Reply:
If you see a better way to refactor I am definitely open to it. I don't think the performance here will be that critical. 
_Probably_ better served as a pre cursor before adding functionality here, or alternately as a follow up to the "side-effect-less" design mentioned already